src/cpufreq.c

```diff
@@ -71,67 +230,222 @@ static void cpufreq_submit (int cpu_num, double value)
+				WARNING ("cpufreq: fgets: %s",
+						 sstrerror (errno, errbuf,
+									sizeof (errbuf)));
+				fclose (fp);
+				return (-1);
+			}
+			
+			if (fclose (fp))
+			{
+				char errbuf[1024];
+				WARNING ("cpufreq: fclose: %s",
+						 sstrerror (errno, errbuf,
+									sizeof (errbuf)));
+			}
+						
+			/* You're seeing correctly: The file is reporting kHz values.. */
+			val = atoll (buffer) * 1000;
+			
+			cpufreq_submit_current_value (i, val);
 		}
+	} /* end of if ( !report_distribution ) */
 
-		if (fgets (buffer, 16, fp) == NULL)
-		{
-			char errbuf[1024];
-			WARNING ("cpufreq: fgets: %s",
-					sstrerror (errno, errbuf,
-						sizeof (errbuf)));
-			fclose (fp);
-			return (-1);
-		}
+	else // stats on CPU frequency distribution is used 
+	{
+		char *fields[2];
+		char statind[64];
+		int numfields;
+		long int hertz;
+		double time;
+		size_t j;
+		size_t last_index = 0;
+
+		if ( !report_by_cpu )
+			for (j=0; j < hertz_size; ++j)
+				time_all_cpus[j] = 0;
 
-		if (fclose (fp))
+		for (i=0; i < num_cpu; ++i)
 		{
-			char errbuf[1024];
-			WARNING ("cpufreq: fclose: %s",
-					sstrerror (errno, errbuf,
-						sizeof (errbuf)));
-		}
+			status = ssnprintf (filename, sizeof (filename),
+								"/sys/devices/system/cpu/cpu%zu/cpufreq/stats/time_in_state",
+								i);
+		
+			if ((status < 1) || ((unsigned int)status >= sizeof (filename)))
+			{
+				WARNING("cpufreq plugin: error in ssnprintf");
+				return (-1);
+			}
+			
+			if ((fp = fopen (filename, "r")) == NULL)
+			{
+				// CPU could be just switched off
+				// let's check the next one
+				continue;
+			}
+
+			while ( fgets(buffer, sizeof(buffer), fp) != NULL )
+			{
+				numfields = strsplit (buffer, fields, STATIC_ARRAY_SIZE(fields));
+				
+				if ( numfields != 2 ) // supported format contains 2 fields
+				{
+					// if its an empty line, just ignore line
+					if ( numfields == 0 ) continue;
+
+					WARNING ("cpufreq: wrong number of items on a line in %s", filename);
+					fclose (fp);
+					return (-1);
+				}
 
+				hertz = atol( fields[0] );
+				time = atof( fields[1] ) * 10e-3;
 
-		/* You're seeing correctly: The file is reporting kHz values.. */
-		val = atoll (buffer) * 1000;
+				if ( hertz <= 0 )
+				{
+					WARNING ("cpufreq: something is wrong in %s: hertz = %ld", filename, hertz);
+					WARNING ("cpufreq: line in question %s %s", fields[0], fields[1]);
+					fclose (fp);
+					return (-1);
+				}
 
-		cpufreq_submit (i, val);
+				if ( report_by_cpu )
+				{
+					status = ssnprintf (statind, sizeof (statind),
+										"%zu", i);
+					
+					if ((status < 1) || ((unsigned int)status >= sizeof (statind)))
+					{
+						WARNING("cpufreq plugin: error in ssnprintf");
+						fclose (fp);
+						return (-1);
+					}
+
+					cpufreq_submit_distribution_value( statind, hertz, time );
+				}
+				else
+				{
+					int found = 0;
+
+					if ( hertz_size==1 )
+					{
+						if ( hertz == hertz_all_cpus[0] )
+							time_all_cpus[0] += time;
```

Comment:
You can't add counts / `derive_t` like this. If you want to aggregate all CPUs, you need to convert each individual count to a *rate* first, using `value_to_rate()` from *common.h*.